Note: This notebook assignment consists of Questions 1, 2 & 3

## Explore and cluster the neighborhoods in Toronto

### Question 1

In [1]:
# Import required libraries
import pandas as pd
import numpy as np
import requests

In [2]:
# use wikipedia webpage in order to obtain the data that is in the table of postal codes and store in url variable
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_url = requests.get(url)
wiki_url

<Response [200]>

Above Response 200 result indicates that the connection with wikipedia page is successful

In [3]:
# read the url data into the panda as a list
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

In [4]:
# Check the tables in wikipedia page and type of data stored in wiki_data variable
len(wiki_data), type(wiki_data)

(3, list)

There are 3 tables in wikipedia page

For this assignment, we need the only first table

In [5]:
# get frist table
wiki_data = wiki_data[0]
wiki_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

Now , we need to only process the cells that have an assigned borough. Ignore cells with a Not assigned values

In [6]:
df = wiki_data[wiki_data["Borough"] != "Not assigned"]
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Grouping the records based on Postal Code

In [7]:
df = df.groupby(['Postal Code']).head()
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Now Check number of records where Neighbourhood is "Not assigned"

In [8]:
df.Neighbourhood.str.count("Not assigned").sum()

0

In [9]:
# reset index
df = df.reset_index()
df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

In [10]:
# drop index column
df.drop(['index'], axis = 'columns', inplace = True)
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [11]:
# finally print the number of rows of dataframe
df.shape

(103, 3)

Answer (Question 1): Dataframe has 103 rows and 3 columns

## Solution of Question 2

First Install geocoder if already not installed

In [12]:
# import geocoder
import geocoder

Tried the below approach, ran for 1 hour, then killed it.

In [ ]:
# # initialize your variable to None
# lat_lng_coords = None

# postal_code = 'M3A'

# # loop until you get the coordinates
# while(lat_lng_coords is None):
#   g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
#   lat_lng_coords = g.latlng

# latitude = lat_lng_coords[0]
# longitude = lat_lng_coords[1]

I will use alternative method, as suggested in the assignment, Importing the CSV file from the URL

In [13]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

get the rows and cloumns of wikipedia and csv file

In [14]:
print("The shape of wiki data is: ", df.shape)
print("the shape of csv data is: ", data.shape)

The shape of wiki data is:  (103, 3)
the shape of csv data is:  (103, 3)


As we can see that the dimensions of both are the same, we can try to join on the postal codes to get the required data.

First, check the column types of Postal Code column to join the both files

In [15]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [16]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

Now join the both dataframes using Postal Code column

In [17]:
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [18]:
# check shape of combine_data
combined_data.shape

(103, 5)

Solution of Question 2: There are 103 rows as expected when we do a inner join.

## Question 3

Following the learning from the lesson labs of New York City  we cluster the neighbourhood of NYC, We cluster Toronto based on the similarities of the venues categories using Kmeans clustering and Foursquare API.

In [20]:
# import the library module
from geopy.geocoders import Nominatim

In [21]:
# get the coordinates of Toronto, Ontario
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


Import the folium libray to generate maps to visualize neighborhoods and how they cluster together

In [22]:
import folium

In [24]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

Initializing Foursquare API credentials

In [25]:
CLIENT_ID = 'HJ1CGMHXBBZBWJ2ENCD3M05G52KJ31EOG4YGHCZKUCECBSP3' # my Foursquare ID
CLIENT_SECRET = 'RN0TY203WDYQRUBVVGOBWRYJ1MQF3WHQIOMGCZCPXP5ZCQEW' # my Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: HJ1CGMHXBBZBWJ2ENCD3M05G52KJ31EOG4YGHCZKUCECBSP3
CLIENT_SECRET:RN0TY203WDYQRUBVVGOBWRYJ1MQF3WHQIOMGCZCPXP5ZCQEW


Next, we create a function to get all the venue categories in Toronto

In [26]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Now collecting the venues in Toronto for each Neighbourhood

In [27]:
venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [28]:
venues_in_toronto.shape

(1325, 5)

So we have 1325 records and 5 columns. Checking sample data

In [29]:
venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2  Victoria Village               43.725882               -79.315572   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue         Venue Category  
0         Brookbanks Park                   Park  
1           Variety Store      Food & Drink Shop  
2  Victoria Village Arena           Hockey Arena  
3             Tim Hortons            Coffee Shop  
4               Portugril  Portuguese Restaurant

Let's check the Venues based on Neighbourhood

In [30]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                      Victoria Village   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1312  Mimico NW, The Queensway West, South of Bloor,...   
1313  Mimico NW, The Queensway West, South of Bloor,...   
1314  Mimico NW, The Queensway West, South of Bloor,...   
1315  Mimico NW, The Queensway West, South of Bloor,...   
1316  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude                   Venue  \
0                  43.753259               -79.329656         Brookbanks Park   
1                  43.753259               -79.329656           Variety Store   
2                  43.725882               -79.315572  Victoria Village Arena   
3                  43.725882               -79.315572             Tim Hortons   
4                  43.725882               -79.315572               Portugril   
...                      ...                      ...                     ...   
1312               43.628841               -79.520999        South St. Burger   
1313               43.628841               -79.520999              Wingporium   
1314               43.628841               -79.520999               Dollarama   
1315               43.628841               -79.520999    Artisano Bakery Café   
1316               43.628841               -79.520999          Healthy Planet   

             Venue Category  
0                      Park  
1         Food & Drink Shop  
2              Hockey Arena  
3               Coffee Shop  
4     Portuguese Restaurant  
...                     ...  
1312           Burger Joint  
1313            Wings Joint  
1314         Discount Store  
1315                 Bakery  
1316        Supplement Shop  

[413 rows x 5 columns]

So there are 413 records for each neighbourhood.

Checking for the maximum venue categories

In [31]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                          
Accessories Store                    Lawrence Manor, Lawrence Heights   
Adult Boutique                                   Church and Wellesley   
Airport                                                     Downsview   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Gate        CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Warehouse Store                                      Thorncliffe Park   
Wine Bar                                     Little Portugal, Trinity   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                                     Scarborough Village   
Yoga Studio                            University of Toronto, Harbord   

                    Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                        
Accessories Store                43.718518               -79.464763   
Adult Boutique                   43.665860               -79.383160   
Airport                          43.737473               -79.394420   
Airport Food Court               43.628947               -79.394420   
Airport Gate                     43.628947               -79.394420   
...                                    ...                      ...   
Warehouse Store                  43.705369               -79.349372   
Wine Bar                         43.653206               -79.400049   
Wings Joint                      43.628841               -79.520999   
Women's Store                    43.744734               -79.239476   
Yoga Studio                      43.715383               -79.340923   

                                              Venue  
Venue Category                                       
Accessories Store               Ardene Shoes Outlet  
Adult Boutique                            Seduction  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Gate                                 Gate 8  
...                                             ...  
Warehouse Store                              Costco  
Wine Bar                            Paris Paris Bar  
Wings Joint                              Wingporium  
Women's Store                         Maximum Woman  
Yoga Studio                               YogaSpace  

[238 rows x 4 columns]

There are around 238 different types of Venue Categories.

## Now let's do One Hot encoding the venue Categories

In [32]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Adult Boutique  Airport  Airport Food Court  \
0                     0               0        0                   0   
1                     0               0        0                   0   
2                     0               0        0                   0   
3                     0               0        0                   0   
4                     0               0        0                   0   
...                 ...             ...      ...                 ...   
1320                  0               0        0                   0   
1321                  0               0        0                   0   
1322                  0               0        0                   0   
1323                  0               0        0                   0   
1324                  0               0        0                   0   

      Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0                0               0                0                 0   
1                0               0                0                 0   
2                0               0                0                 0   
3                0               0                0                 0   
4                0               0                0                 0   
...            ...             ...              ...               ...   
1320             0               0                0                 0   
1321             0               0                0                 0   
1322             0               0                0                 0   
1323             0               0                0                 0   
1324             0               0                0                 0   

      American Restaurant  Antique Shop  ...  Train Station  Truck Stop  \
0                       0             0  ...              0           0   
1                       0             0  ...              0           0   
2                       0             0  ...              0           0   
3                       0             0  ...              0           0   
4                       0             0  ...              0           0   
...                   ...           ...  ...            ...         ...   
1320                    0             0  ...              0           0   
1321                    0             0  ...              0           0   
1322                    0             0  ...              0           0   
1323                    0             0  ...              0           0   
1324                    0             0  ...              0           0   

      Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                                 0                 0                      0   
1                                 0                 0                      0   
2                                 0                 0                      0   
3                                 0                 0                      0   
4                                 0                 0                      0   
...                             ...               ...                    ...   
1320                              0                 0                      0   
1321                              0                 0                      0   
1322                              0                 0                      0   
1323                              0                 0                      0   
1324                              0                 0                      0   

      Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                   0         0            0              0            0  
1                   0         0            0              0            0  
2                   0         0            0              0            0  
3                   0         0            0              0            0  
4                   0      

In [33]:
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Accessories Store  Adult Boutique  Airport  \
0         Parkwoods                  0               0        0   
1         Parkwoods                  0               0        0   
2  Victoria Village                  0               0        0   
3  Victoria Village                  0               0        0   
4  Victoria Village                  0               0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Train Station  Truck Stop  \
0                 0                    0  ...              0           0   
1                 0                    0  ...              0           0   
2                 0                    0  ...              0           0   
3                 0                    0  ...              0           0   
4                 0                    0  ...              0           0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                0         0            0              0            0  
1                0         0            0              0            0  
2                0         0            0              0            0  
3                0         0            0              0            0  
4                0         0            0              0            0  

[5 rows x 239 columns]

Now group the Neighbourhoods, calculate the mean venue categories in each Neighbourhood

In [34]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Adult Boutique  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0             0.0      0.0                 0.0           0.0             0.0   
1             0.0      0.0                 0.0           0.0             0.0   
2             0.0      0.0                 0.0           0.0             0.0   
3             0.0      0.0                 0.0           0.0             0.0   
4             0.0      0.0                 0.0           0.0             0.0   

   Airport Service  Airport Terminal  American Restaurant  ...  Train Station  \
0              0.0               0.0             0.000000  ...            0.0   
1              0.0               0.0             0.000000  ...            0.0   
2              0.0               0.0             0.000000  ...            0.0   
3              0.0               0.0             0.000000  ...            0.0   
4              0.0               0.0             0.041667  ...            0.0   

   Truck Stop  Vegetarian / Vegan Restaurant  Video Game Store  \
0         0.0                            0.0               0.0   
1         0.0                            0.0               0.0   
2         0.0                            0.0               0.0   
3         0.0                            0.0               0.0   
4         0.0                            0.0               0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                    0.0              0.0       0.0          0.0   
1                    0.0              0.0       0.0          0.0   
2                    0.0              0.0       0.0          0.0   
3                    0.0              0.0       0.0          0.0   
4                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 239 columns]

Let's make a function to get the top most common venue categories

In [35]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

There are many venue categories, we can take the top 10 to cluster the neighbourhoods

In [36]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                  Bank   
3                                  Bayview Village                  Bank   
4                Bedford Park, Lawrence Manor East        Sandwich Place   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0  Latin American Restaurant        Breakfast Spot          Skating Rink   
1                        Gym          Skating Rink                  Pool   
2                Coffee Shop                 Diner           Pizza Place   
3         Chinese Restaurant   Japanese Restaurant                  Café   
4                Coffee Shop    Italian Restaurant       Thai Restaurant   

  5th Most Common Venue        6th Most Common Venue  \
0         Deli / Bodega  Eastern European Restaurant   
1                   Pub               Sandwich Place   
2        Ice Cream Shop                  Supermarket   
3           Yoga Studio                Deli / Bodega   
4            Hobby Shop                     Pharmacy   

         7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0                    Drugstore            Donut Shop               Dog Run   
1                  Coffee Shop           Gas Station         Garden Center   
2    Middle Eastern Restaurant            Restaurant     Mobile Phone Shop   
3  Eastern European Restaurant             Drugstore            Donut Shop   
4                    Juice Bar            Restaurant               Butcher   

  10th Most Common Venue  
0    Distribution Center  
1              Gastropub  
2          Deli / Bodega  
3                Dog Run  
4      Indian Restaurant

Let's make the model to cluster our Neighbourhoods

In [37]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [38]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

Check the labelling of our model

In [39]:
kmeans.labels_[0:100]

array([4, 0, 0, 0, 0, 4, 0, 0, 4, 4, 1, 0, 4, 0, 0, 0, 4, 0, 0, 0, 4, 4,
       0, 4, 4, 4, 1, 0, 0, 0, 1, 0, 0, 4, 4, 4, 4, 4, 0, 3, 4, 4, 0, 4,
       4, 4, 4, 4, 4, 4, 1, 4, 1, 0, 1, 4, 4, 4, 3, 4, 0, 1, 0, 0, 0, 1,
       4, 4, 0, 4, 4, 4, 0, 0, 4, 4, 0, 0, 0, 0, 4, 4, 4, 0, 0, 0, 0, 1,
       4, 1, 0, 0, 0, 4, 1, 2])

Now add the clustering Label column to the top 10 common venue categories

In [40]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Join toronto_grouped with combined_data on neighbourhood to add latitude & longitude for each neighborhood to prepare it for plotting

In [41]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             1.0                  Park   
1  43.725882 -79.315572             0.0           Coffee Shop   
2  43.654260 -79.360636             0.0           Coffee Shop   
3  43.718518 -79.464763             4.0        Clothing Store   
4  43.662301 -79.389494             0.0           Coffee Shop   

  2nd Most Common Venue   3rd Most Common Venue  4th Most Common Venue  \
0     Food & Drink Shop             Yoga Studio           Dance Studio   
1          Hockey Arena       French Restaurant  Portuguese Restaurant   
2                Bakery                    Park         Breakfast Spot   
3     Accessories Store  Furniture / Home Store            Coffee Shop   
4      Sushi Restaurant             Yoga Studio    Fried Chicken Joint   

         5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0  Eastern European Restaurant             Drugstore            Donut Shop   
1                  Yoga Studio    Dim Sum Restaurant          Dance Studio   
2                          Pub           Yoga Studio            Restaurant   
3           Miscellaneous Shop              Boutique   Sporting Goods Shop   
4               Sandwich Place                  Park             Nightclub   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0               Dog Run   Distribution Center         Discount Store  
1         Deli / Bodega      Department Store           Dessert Shop  
2                  Café        Chocolate Shop     Mexican Restaurant  
3           Event Space         Women's Store  Vietnamese Restaurant  
4    Mexican Restaurant   Japanese Restaurant     Italian Restaurant

Drop all the NaN values to prevent data skew

In [42]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

Now plot the clusters on the map

In [43]:
# import matplotlib library
import matplotlib.cm as cm
import matplotlib.colors as colors

In [45]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=10)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

Now verify each the clusters

Cluster 1

In [46]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1         North York             0.0           Coffee Shop   
2   Downtown Toronto             0.0           Coffee Shop   
4   Downtown Toronto             0.0           Coffee Shop   
7         North York             0.0                   Gym   
8          East York             0.0           Pizza Place   
9   Downtown Toronto             0.0                  Café   
10        North York             0.0                  Park   
13        North York             0.0                   Gym   
15  Downtown Toronto             0.0                  Café   
17         Etobicoke             0.0                  Café   
19      East Toronto             0.0     Health Food Store   
22       Scarborough             0.0           Coffee Shop   
24  Downtown Toronto             0.0           Coffee Shop   
25  Downtown Toronto             0.0         Grocery Store   
28        North York             0.0                  Bank   
30  Downtown Toronto             0.0                  Café   
33        North York             0.0        Clothing Store   
38       Scarborough             0.0           Coffee Shop   
39        North York             0.0                  Bank   
42  Downtown Toronto             0.0                  Café   
43      West Toronto             0.0                  Café   
48  Downtown Toronto             0.0                  Café   
50        North York             0.0                   Gym   
51       Scarborough             0.0          Skating Rink   
54      East Toronto             0.0           Coffee Shop   
55        North York             0.0        Sandwich Place   
58       Scarborough             0.0       College Stadium   
59        North York             0.0      Ramen Restaurant   
70         Etobicoke             0.0           Pizza Place   
72        North York             0.0         Grocery Store   
74   Central Toronto             0.0                  Café   
76       Mississauga             0.0           Coffee Shop   
79   Central Toronto             0.0          Dessert Shop   
80  Downtown Toronto             0.0                  Café   
81      West Toronto             0.0                  Café   
86   Central Toronto             0.0           Coffee Shop   
88         Etobicoke             0.0                   Gym   
93         Etobicoke             0.0           Pizza Place   
96  Downtown Toronto             0.0    Italian Restaurant   
97  Downtown Toronto             0.0                  Café   
99  Downtown Toronto             0.0   Japanese Restaurant   

    2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
1            Hockey Arena     French Restaurant        Portuguese Restaurant   
2                  Bakery                  Park               Breakfast Spot   
4        Sushi Restaurant           Yoga Studio          Fried Chicken Joint   
7              Beer Store   Japanese Restaurant                         Café   
8               Pet Store             Gastropub           Athletics & Sports   
9                 Theater           Coffee Shop               Clothing Store   
10    Japanese Restaurant                Bakery                  Pizza Place   
13             Beer Store   Japanese Restaurant                         Café   
15              Gastropub           Coffee Shop          Japanese Restaurant   
17      Convenience Store              Pharmacy                  Pizza Place   
19                    Pub          Neighborhood                        Trail   
22               Pharmacy          Soccer Field        Korean BBQ Restaurant   
24     Italian Restaurant                  Café                       Office   
25                   Café                  Park                   Baby Store   
28            Coffee Shop                 Diner                  Pizza Place   
30            Coffee Shop            Steakhouse                 Concert Hall   
33            Coffee Shop            Restaurant 

Cluster 2

In [47]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         North York             1.0                  Park   
21              York             1.0                  Park   
35         East York             1.0                  Park   
49        North York             1.0                  Park   
52        North York             1.0                  Park   
64              York             1.0                  Park   
66        North York             1.0                  Park   
68   Central Toronto             1.0                  Park   
83   Central Toronto             1.0                  Park   
85       Scarborough             1.0                  Park   
91  Downtown Toronto             1.0                  Park   

         2nd Most Common Venue 3rd Most Common Venue  \
0            Food & Drink Shop           Yoga Studio   
21                        Pool         Women's Store   
35           Convenience Store          Intersection   
49  Construction & Landscaping                Bakery   
52                 Yoga Studio           Curling Ice   
64           Convenience Store           Yoga Studio   
66           Convenience Store           Yoga Studio   
68               Jewelry Store      Sushi Restaurant   
83                  Restaurant            Playground   
85                  Playground          Intersection   
91                  Playground                 Trail   

          4th Most Common Venue        5th Most Common Venue  \
0                  Dance Studio  Eastern European Restaurant   
21                  Golf Course  Eastern European Restaurant   
35                  Yoga Studio                 Dance Studio   
49                  Yoga Studio                 Dance Studio   
52  Eastern European Restaurant                    Drugstore   
64                 Dance Studio  Eastern European Restaurant   
66                 Dance Studio  Eastern European Restaurant   
68                        Trail                  Yoga Studio   
83                 Tennis Court                 Dessert Shop   
85                  Curling Ice                    Drugstore   
91             Cuban Restaurant                    Drugstore   

          6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                     Drugstore            Donut Shop               Dog Run   
21                   Donut Shop               Dog Run   Distribution Center   
35  Eastern European Restaurant             Drugstore            Donut Shop   
49  Eastern European Restaurant             Drugstore            Donut Shop   
52                   Donut Shop               Dog Run   Distribution Center   
64                    Drugstore            Donut Shop               Dog Run   
66                    Drugstore            Donut Shop               Dog Run   
68                 Dance Studio             Drugstore            Donut Shop   
83                  Curling Ice          Dance Studio         Deli / Bodega   
85                   Donut Shop               Dog Run   Distribution Center   
91                   Donut Shop               Dog Run   Distribution Center   

   9th Most Common Venue 10th Most Common Venue  
0    Distribution Center         Discount Store  
21        Discount Store                  Diner  
35               Dog Run    Distribution Center  
49               Dog Run    Distribution Center  
52        Discount Store                  Diner  
64   Distribution Center         Discount Store  
66   Distribution Center         Discount Store  
68               Dog Run    Distribution Center  
83      Department Store                  Diner  
85        Discount Store                  Diner  
91        Discount Store                  Diner

Cluster 3

In [48]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
45  North York             2.0             Cafeteria           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
45          Dance Studio     Electronics Store  Eastern European Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
45             Drugstore            Donut Shop               Dog Run   

   9th Most Common Venue 10th Most Common Venue  
45   Distribution Center         Discount Store

Cluster 4

In [49]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue   2nd Most Common Venue  \
57   North York             3.0        Baseball Field  Furniture / Home Store   
101   Etobicoke             3.0        Baseball Field             Yoga Studio   

    3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
57            Yoga Studio          Dance Studio            Electronics Store   
101           Event Space     Electronics Store  Eastern European Restaurant   

           6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
57   Eastern European Restaurant             Drugstore            Donut Shop   
101                    Drugstore            Donut Shop               Dog Run   

    9th Most Common Venue 10th Most Common Venue  
57                Dog Run    Distribution Center  
101   Distribution Center         Discount Store

Cluster 5

In [50]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels      1st Most Common Venue  \
3          North York             4.0             Clothing Store   
6         Scarborough             4.0       Fast Food Restaurant   
12        Scarborough             4.0                        Bar   
14          East York             4.0         Athletics & Sports   
16               York             4.0                      Field   
18        Scarborough             4.0          Electronics Store   
20   Downtown Toronto             4.0               Cocktail Bar   
23          East York             4.0                Coffee Shop   
26        Scarborough             4.0         Athletics & Sports   
27         North York             4.0                Golf Course   
29          East York             4.0          Indian Restaurant   
31       West Toronto             4.0                     Bakery   
32        Scarborough             4.0              Women's Store   
34         North York             4.0         Miscellaneous Shop   
36   Downtown Toronto             4.0                       Café   
37       West Toronto             4.0                        Bar   
40         North York             4.0              Grocery Store   
41       East Toronto             4.0           Greek Restaurant   
44        Scarborough             4.0                   Bus Line   
46         North York             4.0              Grocery Store   
47       East Toronto             4.0             Sandwich Place   
53         North York             4.0              Grocery Store   
56               York             4.0       Fast Food Restaurant   
60         North York             4.0              Grocery Store   
61    Central Toronto             4.0                     Lawyer   
62    Central Toronto             4.0               Home Service   
63               York             4.0              Grocery Store   
65        Scarborough             4.0          Indian Restaurant   
67    Central Toronto             4.0                       Park   
69       West Toronto             4.0         Mexican Restaurant   
71        Scarborough             4.0  Middle Eastern Restaurant   
73    Central Toronto             4.0             Clothing Store   
75       West Toronto             4.0             Breakfast Spot   
77          Etobicoke             4.0                       Park   
78        Scarborough             4.0                     Lounge   
82        Scarborough             4.0                Pizza Place   
84   Downtown Toronto             4.0                       Café   
87   Downtown Toronto             4.0            Airport Service   
89          Etobicoke             4.0              Grocery Store   
90        Scarborough             4.0       Fast Food Restaurant   
92   Downtown Toronto             4.0         Seafood Restaurant   
94          Etobicoke             4.0                  Drugstore   
98          Etobicoke             4.0                      River   
100      East Toronto             4.0         Light Rail Station   
102         Etobicoke             4.0                        Gym   

          2nd Most Common Venue     3rd Most Common Venue  \
3             Accessories Store    Furniture / Home Store   
6                   Yoga Studio               Event Space   
12   Construction & Landscaping               Yoga Studio   
14                         Park              Intersection   
16                      Dog Run                     Trail   
18                   Restaurant            Breakfast Spot   
20               Farmers Market                  Beer Bar   
23       Furniture / Home Store       Sporting Goods Shop   
26                  Gas Station           Thai Restaurant   
27                      Dog Run  Mediterranean Restaurant   
29                  Yoga Studio             Grocery Store   
31                     Pharmacy                      Café   
32                   Playground          Business Service   
34         Caribbean Restaurant

We have successfully cluster Toronto neighbourhood based on venue categories!